src/ApplicationConfigInjectionTrait.php

```diff
@@ -0,0 +1,306 @@
+            }
+
+            return $item;
+        };
+    }
+
+    /**
+     * Create the collection mapping function.
+     *
+     * Returns a callable with the following signature:
+     *
+     * <code>
+     * function (array|string $item) : array
+     * </code>
+     *
+     * When it encounters one of the self::*_MIDDLEWARE constants, it passes
+     * the value to the `createPipelineMapper()` callback to create a spec
+     * that uses the return value as pipeline middleware.
+     *
+     * If the 'middleware' value is an array, it uses the `createPipelineMapper()`
+     * callback as an array mapper in order to ensure the self::*_MIDDLEWARE
+     * are injected correctly.
+     *
+     * If the 'middleware' value is missing, or not viable as middleware, it
+     * raises an exception, to ensure the pipeline is built correctly.
+     *
+     * @return callable
+     */
+    private function createCollectionMapper()
+    {
+        $pipelineMap = $this->createPipelineMapper();
+        $appMiddlewares = [
+            Application::ROUTING_MIDDLEWARE,
+            Application::DISPATCH_MIDDLEWARE,
+        ];
+
+        return function ($item) use ($pipelineMap, $appMiddlewares) {
+            if (in_array($item, $appMiddlewares, true)) {
+                return ['middleware' => $pipelineMap($item)];
+            }
+
+            if (! is_array($item) || ! array_key_exists('middleware', $item)) {
+                throw new Container\Exception\InvalidArgumentException(sprintf(
+                    'Invalid pipeline specification received; must be an array containing a middleware '
+                    . 'key, or one of the Application::*_MIDDLEWARE constants; received %s',
+                    is_object($item) ? get_class($item) : gettype($item)
+                ));
+            }
+
+            if (! is_callable($item['middleware']) && is_array($item['middleware'])) {
+                $item['middleware'] = array_map($pipelineMap, $item['middleware']);
+            }
+
+            return $item;
+        };
+    }
+
+    /**
+     * Create reducer function that will reduce an array to a priority queue.
+     *
+     * Creates and returns a function with the signature:
+     *
+     * <code>
+     * function (SplQueue $queue, array $item) : SplQueue
+     * </code>
+     *
+     * The function is useful to reduce an array of pipeline middleware to a
+     * priority queue.
+     *
+     * @return callable
+     */
+    private function createPriorityQueueReducer()
+    {
+        // $serial is used to ensure that items of the same priority are enqueued
+        // in the order in which they are inserted.
+        $serial = PHP_INT_MAX;
+        return function ($queue, $item) use (&$serial) {
+            $priority = isset($item['priority']) && is_int($item['priority'])
+                ? $item['priority']
+                : 1;
+            $queue->insert($item, [$priority, $serial--]);
```

Comment:
We discuss before `++` operator and here it could be confusing.
I'd prefer to have here just `$serial` and then in new line `$serial--` (or `$serial -= 1`).
Ofc this is post decrementation so value is returned first and then value is changed.